resources/virt.go

```diff
@@ -490,24 +532,306 @@ func (obj *VirtRes) CheckApply(apply bool) (bool, error) {
+	if obj.restartScheduled && obj.RestartOnDiverge != "ignore" { // "ignore", "ifneeded", "error"
+		restart = true
+	}
+	if !apply {
+		restart = false
+	}
+
+	var checkOK = true
+
+	dom, err := obj.conn.LookupDomainByName(obj.GetName())
+	if err == nil {
+		// pass
+	} else if isNotFound(err) {
+		// domain not found
+		if obj.absent {
+			// we can ignore the restart var since we're not running
+			return true, nil
+		}
+
+		if !apply {
+			return false, nil
+		}
+
+		var c = true
+		dom, c, err = obj.domainCreate() // create the domain
+		if err != nil {
+			return false, errwrap.Wrapf(err, "domainCreate failed")
+		} else if !c {
+			checkOK = false
+		}
+
+	} else {
+		return false, errwrap.Wrapf(err, "LookupDomainByName failed")
+	}
+	defer dom.Free() // the Free() for two possible domain objects above
+	// domain now exists
+
+	isPersistent, err := dom.IsPersistent()
+	if err != nil {
+		// we don't know if the state is ok
+		return false, errwrap.Wrapf(err, "domain.IsPersistent failed")
+	}
+	// check for persistence
+	if isPersistent == obj.Transient { // if they're different!
+		if !apply {
+			return false, nil
+		}
+		if isPersistent {
+			if err := dom.Undefine(); err != nil {
+				return false, errwrap.Wrapf(err, "domain.Undefine failed")
+			}
+			log.Printf("%s[%s]: Domain undefined", obj.Kind(), obj.GetName())
+		} else {
+			domXML, err := dom.GetXMLDesc(libvirt.DOMAIN_XML_INACTIVE)
+			if err != nil {
+				return false, errwrap.Wrapf(err, "domain.GetXMLDesc failed")
+			}
+			if _, err = obj.conn.DomainDefineXML(domXML); err != nil {
+				return false, errwrap.Wrapf(err, "conn.DomainDefineXML failed")
+			}
+			log.Printf("%s[%s]: Domain defined", obj.Kind(), obj.GetName())
+		}
+		checkOK = false
+	}
+
+	// shutdown here and let the stateCheckApply fix things up...
+	// TODO: i think this is the most straight forward process...
+	if !obj.absent && restart {
+		if c, err := obj.domainShutdownSync(apply, dom); err != nil {
+			return false, errwrap.Wrapf(err, "domainShutdownSync failed")
+		} else if !c {
+			checkOK = false
+			restart = false // clear the restart requirement...
+		}
+	}
+
+	// FIXME: is doing this early check (therefore twice total) a good idea?
+	// run additional pre-emptive attr change checks here for hotplug stuff!
+	if !obj.absent {
+		if c, err := obj.attrCheckApply(apply, dom); err != nil {
+			return false, errwrap.Wrapf(err, "early attrCheckApply failed")
+		} else if !c {
+			checkOK = false
+		}
+	}
+	// TODO: do we need to run again below after we've booted up the domain?
+
+	// apply correct machine state, eg: startup/shutoff/pause as needed
+	if c, err := obj.stateCheckApply(apply, dom); err != nil {
+		return false, errwrap.Wrapf(err, "stateCheckApply failed")
+	} else if !c {
+		checkOK = false
+	}
+
+	// FIXME: should we wait to ensure machine is booted before continuing?
+	// it may be useful to wait for guest agent to hotplug some ram or cpu!
+
 	if !apply {
 		return false, nil
```

Comment:
I changed this in https://github.com/purpleidea/mgmt/commit/c480bd94db5147fec09a336fb67eac5333091524 Thanks for the comment!